docs/Client.md

```diff
@@ -0,0 +1,1011 @@
+
+Arguments:
+
+* **options** `PipelineOptions`
+* **handler** `(data: PipelineHandlerData) => stream.Readable`
+
+Returns: `stream.Duplex`
+
+#### Parameter: PipelineOptions
+
+Extends: [`RequestOptions`](#parameter-requestoptions)
+
+* **objectMode** `boolean` (optional) - Default: `false` - Set to `true` if the `handler` will return an object stream.
+
+#### Parameter: PipelineHandlerData
+
+* **statusCode** `number`
+* **headers** `IncomingHttpHeaders`
+* **opaque** `unknown`
+* **body** `stream.Readable`
+
+#### Example 1 - Pipeline Echo
+
+```js
+'use strict'
+const { Readable, Writable, PassThrough, pipeline } = require('stream')
+const { createServer } = require('http')
+const { Client } = require('undici')
+
+
+const server = createServer((request, response) => {
+  request.pipe(response)
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  let res = ''
+
+  pipeline(
+    new Readable({
+      read () {
+        this.push(Buffer.from('undici'))
+        this.push(null)
+      }
+    }),
+    client.pipeline({
+      path: '/',
+      method: 'GET'
+    }, ({ statusCode, headers, body }) => {
+      console.log(`response received ${statusCode}`)
+      console.log('headers', headers)
+      return pipeline(body, new PassThrough(), () => {})
+    }),
+    new Writable({
+      write (chunk, _, callback) {
+        res += chunk.toString()
+        callback()
+      },
+      final (callback) {
+        console.log(`Response pipelined to writable: ${res}`)
+        callback()
+      }
+    }),
+    error => {
+      if (error) {
+        console.error(error)
+      }
+
+      client.close()
+      server.close()
+    }
+  )
+})
+```
+
+### `Client.request(options [, callback])`
+
+Performs a HTTP request.
+
+Non-idempotent requests will not be pipelined in order
+to avoid indirect failures.
+
+Idempotent requests will be automatically retried if
+they fail due to indirect failure from the request
+at the head of the pipeline. This does not apply to
+idempotent requests with a stream request body.
+
+Arguments:
+
+* **options** `RequestOptions`
+* **callback** `(error: Error | null, data: ResponseData) => void` (optional)
+
+Returns: `void | Promise<ResponseData>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Parameter: `RequestOptions`
+
+Extends: [`DispatchOptions`](#parameter-dispatchoptions)
+
+* **opaque** `unknown` (optional) - Default: `null` - Used for passing through context to `ResponseData`
+* **signal** `AbortSignal | events.EventEmitter | null` (optional) - Default: `null`
+
+The `RequestOptions.method` property should not be value `'CONNECT'`.
+
+#### Parameter: `ResponseData`
+
+* **statusCode** `number`
+* **headers** `http.IncomingHttpHeaders`
+* **body** `stream.Readable`
+* **trailers** `Record<string, string>` - This object starts out
```

Comment:
I think `Record` means immutable, no? In this case that is not true.

Reply:
I do not believe this is the case. AFAIK this is the recommended way to represent an object in TS. It declares the type of the keys and values so that users cannot do something like: `trailers[99] = ['a', 1, true]`

Reply:
Ah, I might be confusing this with a ecmascript proposal which defines `Record` as immutable.